File: backend/models/Stock.js
const mongoose = require('mongoose');

const StockSchema = new mongoose.Schema({
    ticker: {
        type: String,
        required: true,
        unique: true,
        uppercase: true
    },
    name: {
        type: String,
        required: false
    },
    price: {
        type: Number,
        default: 0
    },
    arabic_stock_fair_value: Number,
    arabic_stock_analyzers_fair_value: Number,
    arabic_stock_getter: String,
    lastUpdated: {
        type: Date,
        default: Date.now
    },

    // --- Main Scores ---
    bf_score: { type: Number, default: 0 },
    fundamental_score: { type: Number, default: 0 },
    technical_score: { type: Number, default: 0 },
    rfp_score: { type: Number, default: 0 },
    rsp_score: { type: Number, default: 0 },
    arabstock_score: { type: Number, default: 0 },
    total_score: { type: Number, default: 0 },

    // --- Trial Scores ---
    trial_bf_score: { type: Number, default: 0 },
    trial_fundamental_score: { type: Number, default: 0 },
    trial_technical_score: { type: Number, default: 0 },
    trial_rfp_score: { type: Number, default: 0 },
    trial_rsp_score: { type: Number, default: 0 },
    trial_arabstock_score: { type: Number, default: 0 },
    trial_total_score: { type: Number, default: 0 }
});

// Index for high-performance sorting and retrieval
StockSchema.index({ total_score: -1, ticker: 1 });

module.exports = mongoose.model('Stock', StockSchema);

File: frontend/lib/models/stock.dart
class Stock {
  final String ticker;
  final String? name;
  final double price;
  final double? arabicStockFairValue;
  final double? arabicStockAnalyzersFairValue;
  final String? arabicStockGetter;
  final DateTime lastUpdated;

  Stock({
    required this.ticker,
    this.name,
    required this.price,
    this.arabicStockFairValue,
    this.arabicStockAnalyzersFairValue,
    this.arabicStockGetter,
    required this.lastUpdated,
  });

  factory Stock.fromJson(Map<String, dynamic> json) {
    return Stock(
      ticker: json['ticker'] ?? 'UNKNOWN',
      name: json['name'],
      price: json['price'] != null ? (json['price'] as num).toDouble() : 0.0,
      arabicStockFairValue: json['arabic_stock_fair_value'] != null 
          ? (json['arabic_stock_fair_value'] as num).toDouble() 
          : null,
      arabicStockAnalyzersFairValue: json['arabic_stock_analyzers_fair_value'] != null 
          ? (json['arabic_stock_analyzers_fair_value'] as num).toDouble() 
          : null,
      arabicStockGetter: json['arabic_stock_getter'],
      lastUpdated: json['lastUpdated'] != null 
          ? DateTime.parse(json['lastUpdated']) 
          : DateTime.now(),
    );
  }
}

class ArabicStockMatch {
  final String title;
  final String link;

  ArabicStockMatch({required this.title, required this.link});

  factory ArabicStockMatch.fromJson(Map<String, dynamic> json) {
    return ArabicStockMatch(
      title: json['title'],
      link: json['link'],
    );
  }
}
